### ISSUE TITLE
max operation not using zmm registers on AVX512

### ISSUE THREAD

**Issue by nlw0** [eschnett/SIMD.jl#47]
I'm trying out SIMD.jl on an AVX512 machine, if I add two 16xFloat32 vectors I can see two zmm registers being used.

```
julia> aa=[Vec(tuple(rand(Float32, 16)...)) for i in 1:8]
8-element Array{Vec{16,Float32},1}:
 <16 x Float32>[0.99203587, 0.19657528, 0.07043815, 0.97100663, 0.7849779, 0.05285406, 0.8422512, 0.58236885, 0.07897723, 0.25636375, 0.97477055, 0.6738832, 0.83993757, 0.019398332, 0.28334594, 0.6181774]
...
julia> @code_native aa[1] + aa[2]
        .text
; ┌ @ SIMD.jl:1054 within `+'
; │┌ @ SIMD.jl:580 within `llvmwrap' @ SIMD.jl:580
; ││┌ @ SIMD.jl:1054 within `macro expansion'
        vmovups (%rsi), %zmm0
        vaddps  (%rdx), %zmm0, %zmm0
; │└└
        vmovaps %zmm0, (%rdi)
        movq    %rdi, %rax
        vzeroupper
        retq
        nopl    (%rax)
; └
```

With the `max` function, though, it seems to fall back to xmm registers

```
julia> @code_native max(aa[1], aa[2])
        .text
; ┌ @ SIMD.jl:1054 within `max'
; │┌ @ SIMD.jl:580 within `llvmwrap' @ SIMD.jl:580
; ││┌ @ SIMD.jl:1054 within `macro expansion'
        vmovups (%rsi), %zmm0
        vmovups (%rdx), %zmm1
        vextractf32x4   $3, %zmm0, %xmm2
        vextractf32x4   $3, %zmm1, %xmm3
        vmaxss  %xmm2, %xmm3, %xmm4
        vcmpunordss     %xmm2, %xmm2, %k1
        vmovss  %xmm3, %xmm0, %xmm4 {%k1}
        vmovshdup       %xmm2, %xmm5    # xmm5 = xmm2[1,1,3,3]
        vmovshdup       %xmm3, %xmm6    # xmm6 = xmm3[1,1,3,3]
        vmaxss  %xmm5, %xmm6, %xmm7
        vcmpunordss     %xmm5, %xmm5, %k1
        vmovss  %xmm6, %xmm0, %xmm7 {%k1}
        vinsertps       $16, %xmm7, %xmm4, %xmm4 # xmm4 = xmm4[0],xmm7[0],xmm4[2,3]
        vpermilpd       $1, %xmm2, %xmm5 # xmm5 = xmm2[1,0]
        vpermilpd       $1, %xmm3, %xmm6 # xmm6 = xmm3[1,0]
        vmaxss  %xmm5, %xmm6, %xmm7
        vcmpunordss     %xmm5, %xmm5, %k1
        vmovss  %xmm6, %xmm0, %xmm7 {%k1}
        vinsertps       $32, %xmm7, %xmm4, %xmm4 # xmm4 = xmm4[0,1],xmm7[0],xmm4[3]
        vpermilps       $231, %xmm2, %xmm2 # xmm2 = xmm2[3,1,2,3]
        vpermilps       $231, %xmm3, %xmm3 # xmm3 = xmm3[3,1,2,3]
        vmaxss  %xmm2, %xmm3, %xmm5
...
```

This is unexpected since on my AVX2 machine it would at worst split the operation into ymm registers. Could this be a bug or some missing configuration for me?

**Comment by eschnett** [comment]
Confirmed.

`@code_llvm` shows that the correct intrinsic `maxnum` is used by Julia:
```llvm
julia> @code_llvm max(Vec{16, Float32}(1), Vec{16, Float32}(1))

;  @ /home/eschnetter/.julia/packages/SIMD/5ugK9/src/SIMD.jl:1012 within `max'
define void @julia_max_12430({ <16 x float> }* noalias nocapture sret, { <16 x float> } addrspace(11)* nocapture nonnull readonly dereferenceable(64), { <16 x float> } addrspace(11)* nocapture nonnull readonly dereferenceable(64)) {
top:
; ┌ @ /home/eschnetter/.julia/packages/SIMD/5ugK9/src/SIMD.jl:538 within `llvmwrap' @ /home/eschnetter/.julia/packages/SIMD/5ugK9/src/SIMD.jl:538
; │┌ @ /home/eschnetter/.julia/packages/SIMD/5ugK9/src/SIMD.jl:557 within `macro expansion'
; ││┌ @ sysimg.jl:18 within `getproperty'
     %3 = getelementptr inbounds { <16 x float> }, { <16 x float> } addrspace(11)* %1, i64 0, i32 0
     %4 = getelementptr inbounds { <16 x float> }, { <16 x float> } addrspace(11)* %2, i64 0, i32 0
; ││└
    %5 = load <16 x float>, <16 x float> addrspace(11)* %3, align 16
    %6 = load <16 x float>, <16 x float> addrspace(11)* %4, align 16
    %res.i = call <16 x float> @llvm.maxnum.v16f32(<16 x float> %5, <16 x float> %6)
; └└
  %.sroa.0.0..sroa_idx = getelementptr inbounds { <16 x float> }, { <16 x float> }* %0, i64 0, i32 0
  store <16 x float> %res.i, <16 x float>* %.sroa.0.0..sroa_idx, align 64
  ret void
}
```

This is then apparently handled catastrophically badly by LLVM.

I tested this with Julia 1.1, which uses LLVM 6.0. LLVM 7.0 is current, with LLVM 8.0 about to be released. I wonder whether this is already corrected in newer versions?

There seems to be a work-around possible:
```julia
julia> @code_native vifelse(f>=g, f, g)
        .text
; ┌ @ SIMD.jl:892 within `vifelse'
; │┌ @ SIMD.jl:892 within `macro expansion'
        vmovdqa (%rsi), %xmm0
        vmovups (%rdx), %zmm1
        movabsq $23455830699248, %rax   # imm = 0x15553C74ECF0
        vptestnmb       (%rax), %xmm0, %k1
        vmovups (%rcx), %zmm1 {%k1}
        vmovaps %zmm1, (%rdi)
        movq    %rdi, %rax
        vzeroupper
        retq
        nopl    (%rax)
; └└
```

If handling nans correctly is not important to you (there are two different, competing standards for handling nans in `max` – one says to return the other argument in this case, the other says to return nan), then we can add a work-around to SIMD that uses this alternative implementation of `max` on this architecture for this LLVM version.

**Comment by nlw0** [comment]
I was going to try rebuilding Julia with a new LLVM since I'm not even sure what I used. Do you think it might actually be an LLVM bug then? At least that should make it easier for me to investigate...

I just made my project public if you want to look (https://github.com/nlw0/ChipSort.jl). I am running sorting networks, and supposedly it's important that we actually use non-branching max and min operators to get the very best possible performance, because it's a lot of them in sequence. So this workaround might not be completely satisfactory, although if it's not splitting the zmm registers it helps already. Anyway, thanks for checking it out!

**Comment by KristofferC** [comment]
Godbolt is pretty good for playing around with LLVM and see what it generates, e.g. https://godbolt.org/z/aA9e99

**Comment by eschnett** [comment]
Nice! I didn't know Godbolt did LLVM input.

If you change the architecture to `skylake-avx512`, then LLVM falls back from `ymm` to `xmm`, as described above. That is with the trunk of clang, so a newer version won't help.

The work-around I described doesn't use branching, it uses the `k` condition registers. The complete code for `max` looks like this:
```julia
julia> max1(x,y) = vifelse(x>=y, x, y)
max1 (generic function with 1 method)

julia> @code_native max1(Vec{32, Float32}(1), Vec{32, Float32}(1))
        .text
; ┌ @ REPL[4]:1 within `max1'
; │┌ @ SIMD.jl:862 within `>='
; ││┌ @ SIMD.jl:593 within `llvmwrap'
; │││┌ @ REPL[4]:1 within `macro expansion'
        vmovups (%rsi), %zmm0
        vmovups 64(%rsi), %zmm1
        vmovups (%rdx), %zmm2
        vmovups 64(%rdx), %zmm3
; │└└└
; │┌ @ SIMD.jl:892 within `vifelse'
; ││┌ @ SIMD.jl:915 within `macro expansion'
        vcmpleps        %zmm1, %zmm3, %k1
        vcmpleps        %zmm0, %zmm2, %k2
        vmovaps %zmm0, %zmm2 {%k2}
        vmovaps %zmm1, %zmm3 {%k1}
; │└└
        vmovaps %zmm3, 64(%rdi)
        vmovaps %zmm2, (%rdi)
        movq    %rdi, %rax
        vzeroupper
        retq
        nopl    (%rax,%rax)
; └
```
where only the middle 4 instructions (marked `SIMD.JL:915`) are relevant; the others are prologue and epilogue and will go away when the function is inlined.

As I mentioned above, this probably doesn't handle `nan` correctly. If that's not important to you, then I can implement the work-around.

Someone should also file a bug report with LLVM, pointing to the Godbolt example.

**Comment by eschnett** [comment]
See https://bugs.llvm.org/show_bug.cgi?id=40984 .

**Comment by nlw0** [comment]
All right, I suppose this workaround might be a great idea then. And I am not concerned with NaNs. By the way, do you think it might make sense to have a special `minmax` function that computes both at the same time?

Hopefully, this will be enough to run my benchmark and see a better performance on a Xeon server than on my notebook!

**Comment by nlw0** [comment]
They have a patch already, amazing! Do you think it is still justifiable to have the workaround? I have no idea how hard it is to use bleeding-edge LLVM in practice.

**Comment by eschnett** [comment]
Julia applies a number of patches to the LLVM libraries it uses. I assume that using a radically newer version of LLVM will lead to problems. It is probably easier to see which patch resolved the issue, and add it (modified if necessary) to Julia's set of LLVM patches. See the `deps/patches` directory in the Julia source code. If you open a bug report in Julia itself, then this patch might even make it into Julia 1.1.1.

**Comment by nlw0** [comment]
Excellent, I'll try that. One doubt I also have is does it all work the same for integer types too? That is actually what I should have been concentrating on from the start.

**Comment by eschnett** [comment]
Integer types are already working fine.

**Comment by KristofferC** [comment]
Note that the fix for this apparently didn't go into LLVM 8 upstream but the patch we bundle with Julia to fix it had a note that said to remove it for LLVM 8. So Julia 1.4 is once again back to the bad codegen for this case. It would probably been good to have added a regression test with the patch PR to make sure that the feature worked when the patch was removed.

Anyway, the fix is in LLVM 9 which is in Julia 1.5 so not much to do until that.

**Comment by eschnett** [comment]
Julia 1.4 hasn't been released yet. We could push for the patch being included at least in Julia 1.4.1.

**Comment by KristofferC** [comment]
Julia 1.4 RC2 is just about to be released and will be promoted to the real release unless it is buggy. If someone makes the work to put the patch in and we do another 1.4 RC I guess it can go into 1.4 though.
